### PR TITLE
feat: Added `transform` to `Rect`

### PR DESCRIPTION
**Pull request by wolfenrain**
# Description

Noticed there wasn't a method for transform rects. This method is kinda biased as it will transform from "center".

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the
relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples`.

## Breaking Change

Does your PR require Flame users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change. (Indicate it in the [Conventional Commit] prefix with a `!`,
  e.g. `feat!:`, `fix!:`).
- [x] No, this is *not* a breaking change.

## Related Issues

*Replace this paragraph with a list of issues related to this PR from the [issue database].
Indicate which of these issues are resolved or fixed by this PR, i.e. Fixes #xxxx*

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md
[Conventional Commit]: https://conventionalcommits.org

### CODE REVIEW

File: packages/flame/lib/src/extensions/matrix4.dart [hunk 1]
```diff
@@ -78,6 +80,13 @@ extension Matrix4Extension on Matrix4 {
     return transform2(out);
   }
 
+  /// Transform [rect] of type [Rect] using the transformation defined by this.
+  Rect transformRect(Rect rect) {
```
Comment (spydon) [inline packages/flame/lib/src/extensions/matrix4.dart:84]: Maybe this could take an `Anchor` too?

File: packages/flame/lib/src/extensions/matrix4.dart [hunk 2]
```diff
@@ -78,6 +80,13 @@ extension Matrix4Extension on Matrix4 {
     return transform2(out);
   }
 
+  /// Transform [rect] of type [Rect] using the transformation defined by this.
```
Comment (st-pasha) [inline packages/flame/lib/src/extensions/matrix4.dart:83]: This method works only with translation transforms and non-negative scale transforms, right?

Comment (wolfenrain) [reply packages/flame/lib/src/extensions/matrix4.dart:83]: Correct, should add that to the dartdoc

File: packages/flame/lib/src/extensions/matrix4.dart (hunk 1 above)
Comment (spydon) [inline packages/flame/lib/src/extensions/matrix4.dart:84]: Maybe this live in a `RectExtension` and take a transform as the argument instead? I think that feels a bit more natural.

Comment (wolfenrain) [reply packages/flame/lib/src/extensions/matrix4.dart:84]: Agreed! That makes more sense to me as well

File: packages/flame/lib/src/extensions/matrix4.dart (hunk 2 above)
Comment (wolfenrain) [reply packages/flame/lib/src/extensions/matrix4.dart:83]: If you have any suggestions, I am all ears.

**Review by erickzanardo — APPROVED** [review]

File: packages/flame/lib/src/extensions/rect.dart [hunk 3]
```diff
@@ -52,6 +53,19 @@ extension RectExtension on Rect {
...
   }
 
+  /// Transform Rect using the transformation defined by [matrix].
+  ///
+  /// **Note:** Rotation matrices will increase the size of the [Rect] but they
+  /// will not rotate it as [Rect] does not have any rotational values.
+  ///
+  /// **Note:** Only non-negative scale transforms are allowed, if a negative
+  /// scale is applied it will return a zero-based [Rect].
+  Rect transform(Matrix4 matrix) {
```
Comment (spydon) [inline packages/flame/lib/src/extensions/rect.dart:63]: It should probably be mentioned that it transforms from the center, since no optional anchor was added.

Comment (st-pasha) [reply packages/flame/lib/src/extensions/rect.dart:63]: I don't think anchor would do anything here. If the transform is only limited to translations and scales, then it does not depend on what the anchor is.

File: packages/flame/lib/src/extensions/rect.dart [hunk 4]
```diff
@@ -52,6 +53,19 @@ extension RectExtension on Rect {
     ];
   }
 
+  /// Transform Rect using the transformation defined by [matrix].
+  ///
+  /// **Note:** Rotation matrices will increase the size of the [Rect] but they
+  /// will not rotate it as [Rect] does not have any rotational values.
```
Comment (luanpotter) [inline packages/flame/lib/src/extensions/rect.dart:59]: I wonder if it would be best to validate and throw errors for invalid matrixes.

We could (in the future) also have a version that fully works with any matrix and returns a [Polygon](https://github.com/flame-engine/flame/blob/main/packages/flame/lib/src/geometry/polygon.dart) to represent non-AA rectangles

File: packages/flame/test/extensions/rect_test.dart [hunk 5]
```diff
@@ -36,5 +37,14 @@ void main() {
       expect(r2.position.x, r1.left);
       expect(r2.position.y, r1.top);
     });
+
+    test('test transform', () {
```
Comment (luanpotter) [inline packages/flame/test/extensions/rect_test.dart:41]: should we test the edge cases mentioned on the comments?

**Review by luanpotter — APPROVED** [review]
LGTM with a couple comments

File: packages/flame/lib/src/extensions/rect.dart (hunk 4 above)
Comment (wolfenrain) [reply packages/flame/lib/src/extensions/rect.dart:59]: I could, but not sure how to validate that on a matrix. So if anyone has any pointers, I am all ears!

File: packages/flame/lib/src/extensions/rect.dart (hunk 3 above)
Comment (wolfenrain) [reply packages/flame/lib/src/extensions/rect.dart:63]: Added a comment for it regardless, but yea adding an anchor isn't possible unless we go for polygons in the future

File: packages/flame/lib/src/extensions/rect.dart [hunk 6]
```diff
@@ -52,6 +53,22 @@ extension RectExtension on Rect {
...
+  ///
+  /// **Note:** Only non-negative scale transforms are allowed, if a negative
+  /// scale is applied it will return a zero-based [Rect].
+  ///
+  /// **Note:** The transformation will always happen from the center of the
+  /// `Rect`.
+  Rect transform(Matrix4 matrix) {
+    final topLeft = matrix.transform2(this.topLeft.toVector2());
+    final bottomRight = matrix.transform2(this.bottomRight.toVector2());
```
Comment (spydon) [inline packages/flame/lib/src/extensions/rect.dart:68]: Is it possible to calculate these values without creating new vectors? Since `transform2` also creates a new vector there will be four new vectors for every call to this method.

Comment (wolfenrain) [reply packages/flame/lib/src/extensions/rect.dart:68]: I can do it without creating any vectors at all!

**Review by spydon — APPROVED** [review]
